### PR TITLE
Only allow the master to update the list of nodes in the cluster state

### PR DESCRIPTION
**Pull request by ywelsch**
The cluster state on a node is updated either
- by incoming cluster states that are received from the active master or
- by the node itself when it notices that the master has gone.

In the second case, the node adds the NO_MASTER_BLOCK and removes the current master as active master from its cluster state. In one particular case, it would also update the list of nodes, removing the master node that just failed. In the future, we want a clear separation between actions that can be executed by a master publishing a cluster state and a node locally updating its cluster state when no active master is around.

### CODE REVIEW

**Review by bleskes — APPROVED** [review]
LGTM but let's give this some CI hours before merging...

File: core/src/test/java/org/elasticsearch/cluster/MinimumMasterNodesIT.java [hunk 1]
```diff
@@ -132,7 +132,9 @@ public void testSimpleMinimumMasterNodes() throws Exception {
         });
         state = client().admin().cluster().prepareState().setLocal(true).execute().actionGet().getState();
         assertThat(state.blocks().hasGlobalBlock(DiscoverySettings.NO_MASTER_BLOCK_ID), equalTo(true));
-        assertThat(state.nodes().getSize(), equalTo(1)); // verify that we still see the local node in the cluster state
+        // verify that both nodes still in the cluster state but no master
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/cluster/MinimumMasterNodesIT.java]: Nit: `both nodes still` -> `both nodes are still` and `but no master` -> `but there is no master`

**Review by jasontedor — CHANGES_REQUESTED** [review]
I left a nit. 😄

File: core/src/test/java/org/elasticsearch/cluster/MinimumMasterNodesIT.java (hunk 1 above)
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/cluster/MinimumMasterNodesIT.java]: I've pushed 77fe95b. Hope this is more to your taste 🐇

**Review by jasontedor — APPROVED** [review]
LGTM.

**Comment by ywelsch** [comment]
@bleskes @jasontedor I ran CI in the afternoon and observed one test failing. The test `testSendingShardFailure` fails as follows when using the "network unresponsive" disruption type:

- isolate node_t2 from node_t0 and node_t1 with "network unresponsive" failure type
- send shard failure from node_t2 to node_t0 which is master
- remove disruption and wait for cluster to be healed (= OK)
- wait for shard failed listener to be called (must have either failed or success).

The last step times out with the change above because the NodeConnectionsService doesn't disconnect anymore from the master during the partition. I wonder if the network failure type "network unresponsive" is unrealistic here as it just swallows the request and keeps it swallowed even after the network partition has been healed.

**Comment by ywelsch** [comment]
I've opened #21107 as possible fix.

**Comment by bleskes** [comment]
yes, the unresponsive network partition is not realistic and should probably be renamed to reflect it. I think in this case we can remove it's usage from the test and push this change. We can then discuss the unresponsive network partition and potential changes to shard state actions separately

**Comment by bleskes** [comment]
the last change LGTM2

**Comment by ywelsch** [comment]
thanks for the reviews @bleskes @jasontedor

**Comment by ywelsch** [comment]
backported to 5.2.0

**Pull request merged**